### PR TITLE
Enable predicate push-down optimization by default.

### PR DESCRIPTION
**Pull request by abyss7**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Improvement

Short description (up to few sentences):

Disable push-down to right table in left join, left table in right join, and both tables in full join.

Detailed description:

The current implementation of push-down has a bug in cases like:
`SELECT a, b FROM (SELECT 1 AS a) ANY LEFT JOIN (SELECT 1 AS a, 1 AS b) USING (a) WHERE b = 0`

It can't be fixed on syntax level, so in the future we have to move this optimization deeper into execution stack.

### CODE REVIEW

File: dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp [hunk 1]
```diff
@@ -34,15 +34,26 @@ namespace ErrorCodes
     extern const int UNKNOWN_ELEMENT_IN_AST;
 }
 
-static constexpr auto and_function_name = "and";
+namespace
+{
+
+constexpr auto and_function_name = "and";
```
Comment (4ertus2) [inline dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp]: It's possible to get its name from Functions: NameAnd::name.

File: dbms/src/Parsers/ASTSelectQuery.cpp [hunk 2]
```diff
@@ -216,11 +216,11 @@ static const ASTArrayJoin * getFirstArrayJoin(const ASTSelectQuery & select)
 static const ASTTablesInSelectQueryElement * getFirstTableJoin(const ASTSelectQuery & select)
 {
     if (!select.tables)
-        return {};
+        return nullptr;
```
Comment (4ertus2) [inline dbms/src/Parsers/ASTSelectQuery.cpp]: Do we really need it? I'd prefere empty list to be able to change pointer to object without need to change return statements all over the code.

File: dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp [hunk 3]
```diff
@@ -121,15 +142,71 @@ bool PredicateExpressionsOptimizer::allowPushDown(const ASTSelectQuery * subquer
...
+            /// Skip lambda, tuple and other special functions
             if (function && function->isStateful())
                 return false;
         }
+    }
+
+    const auto * ast_join = ast_select->join();
+
+    /// XXX: code depends on the order and availability of AST children
+    const auto * left_table_expr = ast_join ? ast_select->tables->as<ASTTablesInSelectQuery>()
```
Comment (4ertus2) [inline dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp]: Is it better to set pointers with nullptrs first & make common if (ast_join) for both of them?

File: dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp [hunk 4]
```diff
@@ -121,15 +142,71 @@ bool PredicateExpressionsOptimizer::allowPushDown(const ASTSelectQuery * subquer
...
-        return true;
+        /// Check left side for RIGHT'o'FULL JOIN
+        if (isRightOrFull(ast_join->table_join->as<ASTTableJoin>()->kind) && left_subquery == subquery)
+        {
+            return false;
+        }
     }
 
-    return false;
+    for (const auto & [identifier, prefix] : dependencies)
```
Comment (4ertus2) [inline dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp]: I've spent some time to extract this cycle into own func to be clear that it doesn't use more arguments that function gets. I don't think it's a good idea to mix it with other the code again cause one day it starts to use all the other variables.

File: dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp]: No!

Code should not depend on `Functions/...` (only dependency on `FunctionsFactory` and `IFunction` is allowed).

File: dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp (hunk 4 above)
Comment (abyss7) [reply dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp]: Fixed

File: dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp (hunk 3 above)
Comment (abyss7) [reply dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp]: Refactored

**Comment by abyss7** [comment]
@alexey-milovidov I don't see any differences in the PVS report, that you mentioned earlier in Telegram Chat. Can you point out what new errors/warnings were introduced in this PR?

File: dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp [hunk 5]
```diff
@@ -127,15 +148,77 @@ bool PredicateExpressionsOptimizer::allowPushDown(const ASTSelectQuery * subquer
...
 
-            /// Skip lambda、tuple and other special functions
+            /// Skip lambda, tuple and other special functions
             if (function && function->isStateful())
                 return false;
         }
+    }
+
+    const auto * ast_join = ast_select->join();
+    const ASTTableExpression * left_table_expr = nullptr, * right_table_expr = nullptr;
```
Comment (KochetovNicolai) [inline dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp]: I think it conflicts with our code style.

**Review by KochetovNicolai — APPROVED** [review]

File: dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp [hunk 6]
```diff
@@ -127,15 +148,92 @@ bool PredicateExpressionsOptimizer::allowPushDown(const ASTSelectQuery * subquer
...
+        /// NOTE: the syntactic way of pushdown has limitations and should be partially disabled in case of JOINs.
+        ///       Let's take a look at the query:
+        ///
+        ///           SELECT a, b FROM (SELECT 1 AS a) ANY LEFT JOIN (SELECT 1 AS a, 1 AS b) USING (a) WHERE b = 0
+        ///
+        ///       The result is empty - without pushdown. But the pushdown tends to modify it in this way:
+        ///
+        ///           SELECT a, b FROM (SELECT 1 AS a) ANY LEFT JOIN (SELECT 1 AS a, 1 AS b WHERE b = 0) USING (a) WHERE b = 0
+        ///
+        ///       That leads to the empty result in the right subquery and changes the whole outcome to (1, 0) or (1, NULL).
```
Comment (zhang2014) [inline dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp]: 👍 Are we going to move the entire predicate optimization onto the execution stack in the future ? and how do we do that ? Maybe I can get on with this work?

**Comment by alexey-milovidov** [comment]
> I don't see any differences in the PVS report, that you mentioned earlier in Telegram Chat. Can you point out what new errors/warnings were introduced in this PR?

There is no difference in PVS report for now.

**Comment by alexey-milovidov** [comment]
00933_ttl_simple in tsan - TODO @alesapin

File: dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp (hunk 6 above)
Comment (alexey-milovidov) [reply dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp]: Yes, we are going to do it after "Processors" will be ready https://github.com/yandex/ClickHouse/pull/4914
And we will be happy to have your help!

**Pull request merged**